### ISSUE TITLE
General Refactoring Support

### ISSUE THREAD

**Issue by arunhallan** [microsoft/pylance-release#587]
Hi,

Are there any plans to offer refactoring support?

For example, when typing a new class name, VS code can automatically generate the class.

And extract class to a different file?

This is one thing that is making me consider PyCharm (although I really really don't want to).

**Comment by judej** [comment]
Yes, we will have _some_ refactoring available soon. Hang tight :)

**Comment by arunhallan** [comment]
> Yes, we will have _some_ refactoring available soon. Hang tight :)

Great! How soon? :)

**Comment by jakebailey** [comment]
The latest insiders today (`"pylance.insidersChannel": "daily"`) includes preview support for code extraction actions.

**Comment by arunhallan** [comment]
Great! Is there any documentation for the actions?

**Comment by jakebailey** [comment]
The refactorings will show up as quick actions in the lightbulb menu or when you hit `Ctrl+.`. There isn't much to document past that.

We're still working on making the core extension's built-in refactorings be disabled when Pylance is running.

**Comment by chinoto** [comment]
Took me a bit to understand why the lightbulb showed up once, but not after that, turns out your selection has to be exactly around the thing to be manipulated.
I haven't used refactoring too much, so I don't know how common this capability is, but with `matklad.rust-analyzer` (for Rust), simply having your cursor inside a string or partially selecting a method chain will allow you to extract the code into either a variable or function.
Probably not hugely useful, but making the refactoring options show up on fuzzier selections makes it more obvious that you could do a refactoring.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
